### ISSUE TITLE
Completely different training results on TF 2.4.1 and 2.0.0

### ISSUE THREAD

**Issue by FabianaFerreira** [tensorflow/tensorflow#49453]
Hi! How are you? Hope y'all good.

I have a very specific scenario and it would be awesome to have your input. I'm working on a project in which I've developed a training script using TF 2.4.1. Just to give you context, I used MobileNetV2 as the base model to feature extraction and just a Dense layer with one neuron, and, with time, it got necessary to have different base models so as to perform a benchmark, so I started using different base models, such as InceptionV3 and VGG16, defining it as a command option while running my script.

However, so as to the others in my research team to run my script, I had to use TF 2.0.0 due to a software limitation on the laboratory machine (CUDA 10 instead of 11), and that's when it started getting weird. I checked on the documentation if anything in the methods that I was using changed, so as to adapt them. The only changed that I saw was that in TF 2.4.1, the `fit` method supported generators, so I changed my script (using TF 2.0.0) to use the `fit_generator` method, as you can see below:
- TF 2.4.1
  ```python
    training = model.fit(
        generator=train_data_generator,
        epochs=EPOCHS_NUMBER,
        validation_data=valid_data_generator,
        callbacks=callbacks,
    )
  ``` 

- TF 2.0.0
  ```python
    training = model.fit_generator(
        generator=train_data_generator,
        epochs=EPOCHS_NUMBER,
        validation_data=valid_data_generator,
        callbacks=callbacks,
    )
  ```

Besides that, I changed nothing in my script, which you can find at the end of this description. So, okay, what's the problem, right? It occurs that the validation loss does not decrease while running the script using TensorFlow 2.0.0. I tried to check the documentation, if there were any breaking changes or something like this, after checking if the data and the seeds were the same. However, I wasn't able to find anything that helped me understand why the validation loss changed so much, the performance was degraded and due to the fact that I'm using early stopping, the training was finishing earlier than expected, resulting in a worse test performance than the one that I got running with TF 2.4.1. For instance, running the same script, initializing the seed with the same value, using the two different versions, I got the following results:

- TF 2.0.0

name |  val_accuracy | val_loss | val_precision | val_recall | test_accuracy | test_loss | test_precision | test_recall
-- | -- | -- | -- | -- | -- | -- | -- | -- |
fold_5 | 80.2899% | 0.442051 | 85.1852% | 73.3333% | 81.2022% | 0.456165 | 93.0514% | 67.3961%
fold_4 |  77.3913% | 0.507347 | 85.6604% | 65.7971% | 83.4973% | 0.412218 | 95.0000% | 70.6783%
fold_3 |  75.9420% | 0.517585 | 78.7781% | 71.0145% | 76.3934% | 0.515756 | 81.6273% | 68.0525%
fold_2 |  74.8191% | 0.595039 | 95.2381% | 52.1739% | 74.7541% | 0.521118 | 95.9350% | 51.6411%
fold_1 | 73.0825% | 0.652833 | 92.5532% | 50.2890% | 72.5683% | 0.590006 | 96.3964% | 46.8271%

For instance, in the fold 5, this is how the loss and the accuracy are evolving with time:
![image](https://user-images.githubusercontent.com/19495917/119069893-57ec6600-b9bd-11eb-9b43-3bd9d9ae5bc1.png)

- TF 2.4.1

name | val_accuracy | val_loss | val_precision | val_recall | test_accuracy | test_loss | test_precision | test_recall
-- | -- | -- | -- | -- | -- | -- | -- | -- |
fold_5 | 96.6667% | 0.096043 | 96.0000% | 97.3913% | 95.5191% | 0.156118 | 92.2764% | 99.3435%
fold_4 | 94.4928% | 0.134758 | 92.5208% | 96.8116% | 96.3934% | 0.145110 | 93.6214% | 99.5624%
fold_3  | 96.0870% | 0.094560 | 96.7647% | 95.3623% | 96.3934% | 0.136126 | 94.1667% | 98.9059%
fold_2 | 97.3951% | 0.101784 | 97.6676% | 97.1014% | 96.6120% | 0.131193 | 94.5607% | 98.9059%
fold_1 | 96.3821% | 0.095308 | 97.3451% | 95.3757% | 96.5027% | 0.118284 | 95.1168% | 98.0306%

In the same fold (hoping that the seed is set correctly), but using the newest TF version, we have the following graphics:
![image](https://user-images.githubusercontent.com/19495917/119070039-a7cb2d00-b9bd-11eb-874e-ebea8ed94050.png)

It's possible to notice in the first plot that it seems that nothing is happening in the validation data. It seems that the model kind of learns something, because the training loss is decreasing, but "nothing" is happening with the validation loss. I honestly tried to search for a lot of things in order to explain this behavior, since it's exactly the same code but wasn't able to find anything useful. I do understand that, depending on the implementation, fluctuations might happen, but it seems to me that the scenario that I presented before is really specific.

Unfortunately, this is a no-go for my project because I do need to run the experiments in the other machine, which has CUDA 10, but I develop the whole application in the computer that I have access to, which is my personal one. 

Given this scenario, can you, please, help me trying to understand what is going on? Am I missing any other change that I have to do in the script? Is this type of behavior expected? I would appreciate any help! It might be worth saying that for me to test it, I installed both CUDAs (10.1 and 11) on my computer and created two different _conda_ environments so I can switch between them easily.

I'm sorry if I'm not in the right channel to ask this, and if it's not, please redirect me to the correct one. Thank you so much!

Below you can find my training script:
```python
from pathlib import Path

import mlflow.tensorflow
from sklearn.model_selection import GroupKFold, StratifiedKFold
from tensorflow.keras.callbacks import Callback, EarlyStopping
from tensorflow.keras.layers import Dense
from tensorflow.keras.metrics import Precision, Recall
from tensorflow.keras.preprocessing.image import ImageDataGenerator

from classes.Evaluator import Evaluator
from utils.args import get_args_train
from utils.inputs import get_inputs_paths_and_targets
from utils.model import build_model, get_preprocess_function
from utils.seed import set_seeds

# Setting the random_state to get reproducible results
seed = set_seeds()

# Constants
ROOT_FOLDER = Path(__file__).resolve().parent.parent
FOLDS_NUMBER = 5
EPOCHS_NUMBER = 100
EPOCHS_WAITING_FOR_IMPROVEMENT = 5

# Gets which input we're going to use
args = get_args_train()
IMAGE_FOLDER_PATH = ROOT_FOLDER / f"crop_result_prop_{args.proportion}"


class TrainingAndValidationMetrics(Callback):
    def on_epoch_end(self, epoch, logs=None):
        mlflow.log_metrics(logs, step=epoch)


# Instantiates GroupKFold class to split into train and test
group_k_fold = GroupKFold(n_splits=FOLDS_NUMBER)

# Configuration dictionary that is going to be used to compile the model
config = {
    "optimizer": "adam",
    "loss": "binary_crossentropy",
    "metrics": ["accuracy", Precision(), Recall()],
}

# Gets the dataframe that are going to be used in the flow_from_dataframe
# method from the ImageDataGenerator class
input_df = get_inputs_paths_and_targets(args.proportion)

idg = ImageDataGenerator(
    fill_mode="nearest",
    preprocessing_function=get_preprocess_function(args.model.lower()),
)

# Using GroupKFold only to guarantee that a group (in this case, the slide)
# will contain data only in the train or the test group
for train_idx, test_idx in group_k_fold.split(
    input_df.input, input_df.target, input_df.slide
):
    train_data = input_df.iloc[train_idx]
    test_data = input_df.iloc[test_idx]

    # Break here is being used to get only the first fold
    break

print("### Testing data distribution ###")
print(f"{test_data.groupby('slide').count()}")

generator_kwargs = {
    "directory": IMAGE_FOLDER_PATH,
    "x_col": "input",
    "y_col": "target",
    "seed": seed,
    "classes": ["0", "1"],
}

# Generator that will be used to evaluate the model
test_data_generator = idg.flow_from_dataframe(
    test_data, class_mode="binary", shuffle=False, **generator_kwargs
)

# Instantiates the Early Stopping that is going to be used
# in the fit method
early_stopping = EarlyStopping(
    monitor="val_loss", patience=EPOCHS_WAITING_FOR_IMPROVEMENT
)

# Defines the list of callbacks that is an argument of the fit
# method
callbacks = [early_stopping, TrainingAndValidationMetrics()]

# K-fold Cross Validation model evaluation
kfold = StratifiedKFold(n_splits=FOLDS_NUMBER, shuffle=True, random_state=seed)

current_fold = 1
# Starts run on mlflow to register metrics (experiment)
mlflow.start_run(
    run_name=f"{args.model.lower()}",
    tags={"data_proportion": args.proportion, "environment": "main"},
)


# Folds iteration
for train_idx, val_idx in kfold.split(train_data.input, train_data.target):

    # Builds the model
    model = build_model(args.model.lower(), [Dense(1, activation="sigmoid")], config)

    # Starts run on mlflow to register metrics (runs)
    with mlflow.start_run(run_name=f"fold_{current_fold}", nested=True):

        fitting_data = train_data.iloc[train_idx]
        val_data = train_data.iloc[val_idx]

        mlflow.log_text(
            f"Training data: \n{fitting_data.groupby('target').count()} \n"
            f"Validation data: \n{val_data.groupby('target').count()} \n"
            f"Data proportion: {args.proportion} \n"
            f" ----- Using Stratified KFold with seed 1 ----- ",
            artifact_file="data_description.txt",
        )

        train_data_generator = idg.flow_from_dataframe(
            fitting_data, class_mode="binary", **generator_kwargs
        )
        valid_data_generator = idg.flow_from_dataframe(
            val_data, class_mode="binary", **generator_kwargs
        )

        training = model.fit(
            train_data_generator,
            epochs=EPOCHS_NUMBER,
            validation_data=valid_data_generator,
            callbacks=callbacks,
        )

        # Logging model
        mlflow.keras.log_model(keras_model=model, artifact_path="model")

        evaluator = Evaluator(model, training, test_data_generator, test_data.target)

        # Classification report
        test_metrics = evaluator.evaluate_model()
        mlflow.log_metrics(test_metrics)

        # Saves the classification report generated from the predict method
        mlflow.log_text(
            evaluator.generate_classification_report(),
            artifact_file="classification_report.txt",
        )

        # Saves the accuracy and the loss per epoch
        mlflow.log_figure(
            evaluator.generate_training_history_image(),
            artifact_file="accuracy_loss_epochs.png",
        )

        # Saves ROC figure
        mlflow.log_figure(evaluator.generate_roc_figure(), artifact_file="roc.png")

    current_fold += 1

mlflow.end_run()
```

**Comment by tilakrayal** [comment]
@FabianaFerreira,

This issue is not a TensorFlow bug or feature request, it is better asked on [StackOverflow](https://stackoverflow.com/questions/tagged/tensorflow). There is a larger community that reads questions there. 

Thanks!

**Comment by FabianaFerreira** [comment]
Thank you so much, @tilakrayal and sorry for posting on the wrong community channel! I'm closing the issue.
